### PR TITLE
Fuseki improvements for access control, servlet context, and configurations with shared datasets.

### LINKED ISSUE #1957: Add Fuseki main option to set the ServletConfig path
### Version

4.9.0

### Feature

The FusekiServer builder has the calls to set the servlet context path but the command line tool doesn't expose it.

It can be convenient to separate the path prefix from the dataset service names.

### Are you interested in contributing a solution yourself?

Yes

### LINKED ISSUE #1958: Add a pseudo user name for "deny all"
### Version

4.9.0

### Feature

At the moment, "deny all" has to be the cryptic:

```
    ja:allowedUsers () ;
```
or
```
    ja:allowedUsers "" ;
```

There isn't a a string way to write it clearly.

Add user "!" which is "deny all" and also can not be combined with other users (similar to "*", all logged in users, and "_" any user request, logged in or not).

### Are you interested in contributing a solution yourself?

Yes

### LINKED ISSUE #1959: Named create datasets objects.
### Version

4.9.0

### Feature

In a Fuseki configuration file, dataset for services are shared if they refer to the same resource in the configuration file. This is done by the Fuseki processing and only applies to datasets defined in a service `fuseki:dataset`. It does not apply in general assemblers and so does not apply in a Fuseki configuration 

Concretely, this shares one dataset between two services:
```
:service1 a fuseki:Service
    fuseki:dataset :database

:service2 a fuseki:Service
    fuseki:dataset :database

:database a ja:MemoryDataset
```
but with an intent to have both RDFS and plain access to the same storage database, this does not work:
```
:service1 a fuseki:Service
    fuseki:dataset :databaseRDFS

:service2 a fuseki:Service
    fuseki:dataset :databaseBase

:datasetRDFS rdf:type ja:DatasetRDFS ;
    ja:dataset :datasetBase ;

:databaseBase a ja:MemoryDataset .
```
and results in two `ja:MemoryDataset`.

It is a limitation of the assembler subsystem. There is no outer "start assembly - finish assembly" to manage a cache over the entire assembly operation, nor a passed around context in the assembler open call. Assemblers are static objects and not created each time an assemble cycle is performed. The `Mode` argument only says whether to share objects, not give a place to manage the sharing of objects.

A proper fix looks like it is an API change to assemblers.

For now, add a dataset-fix : allow datasets to have a name, and visiting a dataset description with a name will create the object once then share it. This is enough for Fuseki, and for other one-time use of assemblers such as dataset descriptions on the command line to SPARQL query.

### Are you interested in contributing a solution yourself?

Yes

### PR DESCRIPTION
**Pull request by afs**
GitHub issue resolved #1957
GitHub issue resolved #1958
GitHub issue resolved #1959

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx, or if in JIRA, JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

### CODE REVIEW

File: jena-arq/src/main/java/org/apache/jena/query/DatasetFactory.java [hunk 1]
```diff
@@ -254,16 +254,48 @@ public static Dataset assemble(String filename, String resourceURI) {
...
+     * @param model The Model
+     * @param resourceURI The URI for the the dataset in the assembler model.
+     * @return Dataset
+     */
+	public static Dataset assemble(Model model, String resourceURI) {
+        Objects.requireNonNull(model, "model can not be null") ;
+        Objects.requireNonNull(resourceURI, "resourceURI can not be null") ;
+        Resource r = model.createResource(resourceURI);
+        return assemble(r);
+	}
```
Comment (kinow) [inline jena-arq/src/main/java/org/apache/jena/query/DatasetFactory.java:268]: The indenting appears to be wrong looking at the GitHub UI diff. Probably tabs & spaces used in the same file, so no need to change I think.

File: jena-fuseki2/jena-fuseki-main/src/test/java/org/apache/jena/fuseki/main/access/TestAuthorized.java [hunk 2]
```diff
@@ -53,8 +55,28 @@ public class TestAuthorized {
...
+        AuthPolicy auth = Auth.policyAllowSpecific("!");
+        assertFalse(auth.isAllowed(null));
+        assertFalse(auth.isAllowed("user1"));
+    }
+
+    @Test public void auth_anyLoggedIn_4() {
+        String level = LogCtl.getLevel(Fuseki.configLog);
+        try {
+            LogCtl.set(Fuseki.configLog, "Error");
+            // "!" anbd anythign else causes a warning.
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-main/src/test/java/org/apache/jena/fuseki/main/access/TestAuthorized.java:68]: s/anbd/and

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java [hunk 3]
```diff
@@ -867,6 +874,25 @@ private static boolean argBoolean(Resource r, Property p, boolean dftValue) {
             }
         }
 
+        private static String argString(Resource r, Property p, String dftValue) {
+            try { GraphUtils.atmostOneProperty(r, p); }
+            catch (NotUniqueException ex) {
+                throw new FusekiConfigException(ex.getMessage());
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java:880]: Is it worth to include the `ex` in the `FusekiConfigException`, to have more context in the error?

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java [hunk 4]
```diff
@@ -867,6 +874,25 @@ private static boolean argBoolean(Resource r, Property p, boolean dftValue) {
...
+            Statement stmt = r.getProperty(p);
+            if ( stmt == null )
+                return dftValue;
+            try {
+                Node n = stmt.getObject().asLiteral().asNode();
+                if ( ! G.isString(n) )
+                    throw new FusekiConfigException("Not a string for '"+p+"' : "+stmt.getObject());
+                return n.getLiteralLexicalForm();
+            } catch (JenaException ex) {
+                throw new FusekiConfigException("Not a string for '"+p+"' : "+stmt.getObject());
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java:891]: Ditto here

File: jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/build/FusekiConfig.java [hunk 5]
```diff
@@ -635,19 +637,24 @@ private static void accEndpoint(Collection<Endpoint> endpoints, Operation operat
 
     public static Dataset getDataset(Resource datasetDesc, DatasetDescriptionMap dsDescMap) {
         // check if this one already built
+        // This is absolute and does not require on NamedDatasetAssembler;
```
Comment (kinow) [inline jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/build/FusekiConfig.java:640]: s/on/the ? s/on/one? or s/does not require/is not required?

**Review by kinow — APPROVED** [review]

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java (hunk 3 above)
Comment (afs) [reply jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/FusekiServer.java:880]: In this case, the original exception will in the function called, not with a deeper stack. There isn't more context. The effect is to change the name of the exception to `FusekiConfiguration`.

File: jena-arq/src/main/java/org/apache/jena/query/DatasetFactory.java (hunk 1 above)
Comment (afs) [reply jena-arq/src/main/java/org/apache/jena/query/DatasetFactory.java]: Thanks.  GH UI is the place when tabs vs spaces isn't a realistic choice. Spaces win.

**Comment by afs** [comment]
Thanks for the comments.

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/cmds/FusekiMain.java [hunk 6]
```diff
@@ -442,6 +448,15 @@ protected void processModulesAndArgs() {
 
         // -- Server setup.
 
+        if ( contains(argPathBase) ) {
+            // Static files.
+            String servletContextPath = getValue(argPathBase);
+
+            if ( servletContextPath.endsWith("/") )
```
Comment (rvesse) [inline jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/cmds/FusekiMain.java:455]: This check shouldn't trigger on a plain `/` as that's a valid value for the root context

File: jena-fuseki2/jena-fuseki-main/src/test/java/org/apache/jena/fuseki/main/TestFusekiServerBuild.java [hunk 7]
```diff
@@ -412,6 +415,26 @@ public class TestFusekiServerBuild {
         } finally { server.stop(); }
     }
```
Comment (rvesse) [inline jena-fuseki2/jena-fuseki-main/src/test/java/org/apache/jena/fuseki/main/TestFusekiServerBuild.java:417]: Would be useful to have an explicit test of setting `/` as the context path, while that is the implicit default someone will explicitly try to set it at some point

**Review by rvesse — CHANGES_REQUESTED** [review]

File: jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/cmds/FusekiMain.java (hunk 6 above)
Comment (afs) [reply jena-fuseki2/jena-fuseki-main/src/main/java/org/apache/jena/fuseki/main/cmds/FusekiMain.java]: Fixed

File: jena-fuseki2/jena-fuseki-main/src/test/java/org/apache/jena/fuseki/main/TestFusekiServerBuild.java (hunk 7 above)
Comment (afs) [reply jena-fuseki2/jena-fuseki-main/src/test/java/org/apache/jena/fuseki/main/TestFusekiServerBuild.java:440]: Test added ... with some sidetracking into clearing up testing (I found some tests passed for the wrong reasons).

**Review by rvesse — APPROVED** [review]

**Pull request merged**